component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java

```diff
@@ -110,7 +112,13 @@ protected void initiateAuthenticationRequest(HttpServletRequest httpServletReque
             throws AuthenticationFailedException {
         try {
             if (authenticationContext.isRetrying()) {
-                String errorPageUrl = IdentityUtil.getServerURL(X509CertificateConstants.ERROR_PAGE, false, false);
+                String errorPageUrl;
+                try {
+                    errorPageUrl = ServiceURLBuilder.create().addPath(X509CertificateConstants.ERROR_PAGE).build()
+                            .getAbsoluteInternalURL();
+                } catch (URLBuilderException e) {
+                    throw new RuntimeException("Error occurred while building URL.", e);
```

Comment:
Any reason to throw a RuntimeException?

Reply:
We have been generally throwing runtime exceptions when building the URLs and it should be the case since we are building the URLs dynamically using parameters from runtime.